orchestrator/orderedoffsets.go

```diff
@@ -0,0 +1,63 @@
+package orchestrator
+
+import (
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"sort"
+)
+
+type MinOrderedOffsets struct {
+	offsets []kafka.TopicPartition
+}
+
+func NewOrderedOffsets() *MinOrderedOffsets {
+	return &MinOrderedOffsets{
+		offsets: make([]kafka.TopicPartition, 0),
+	}
+}
+
+func (o *MinOrderedOffsets) Insert(offset kafka.TopicPartition) {
+	index := sort.Search(len(o.offsets), func(i int) bool { return o.offsets[i].Offset > offset.Offset })
```

Comment:
You can avoid most sorting and searching by changing your approach.

Consider a (ringbuffer or) slice that has an array of work items that are sorted in offset order.  What states can work be in?

It can be:
1) received from Kafka but not started;
2) started but not completed;
3) completed (or timed-out or failed) but not committed; or,
4) committed but not removed (from the buffer).

When a worker completes work, you find the work in the queue.  Since its queue position does not change (with a proper indexing scheme), then you can find it in constant time. So, to update the status of the work, you simple using its index and update its status.

Now, you merely need to walk this buffer from the beginning on each update.  For each completed item, you can commit its offset and effectively remove the item from the buffer. 

If you want to implement timeouts, then you can simply schedule a goroutine to inspect the queue when the shortest timeout expires.  This updates the queue and triggers the above pruning.


----
Aside on ringbuffers:

A ringbuffer is a circular array.  If you limit the number of incomplete work items that you allow to be inflight, then you can pre-allocate the ringbuffer to be a fixed size.  Then, you will not need to allow any heap memory.

If you represent the ringbuffer as a slice, then you can use the index into the slice as the index of an item.  However, this is problematic because it introduces aliasing.  An alternative is to index by an integer than is always increasing. In this way, each item that is placed into the ringbuffer has a unique index.  You can then map the that index back to an index into the buffer using modular arithmetic.

```
buffer[m % len(buffer)] = { partition: x, offset: y}
m++
```
In this example, the address into the buffer is `m`. `m` is effectively never reused (assuming the type of m is sufficiently large, i.e.`int64`, such that overflow is unlikely to occur).

The cool think about using this addressing scheme is that you can also use it even if you allow your ringbuffer to shrink or grow.

Reply:
Great suggestion - the will simplify the implementation significantly. I have some questions regarding the indexing scheme, but we can discuss this later. I'd like to get this version deployed before I do major changes.